runtime/src/accounts_db.rs

```diff
@@ -2150,27 +2766,233 @@ impl AccountsDB {
+    // > MAX_CACHE_SLOTS of them.
+    pub fn force_flush_accounts_cache(&self) {
+        self.flush_accounts_cache(true);
+    }
+
+    pub fn flush_accounts_cache_if_needed(&self) {
+        self.flush_accounts_cache(false);
+    }
+
+    fn flush_accounts_cache(&self, force_flush: bool) {
+        if !force_flush && self.accounts_cache.num_slots() <= MAX_CACHE_SLOTS {
+            return;
+        }
+
+        // Flush all roots
+        let mut flush_roots_elapsed = Measure::start("flush_roots_elapsed");
+        let cached_roots = self.accounts_cache.clear_roots();
+        for root in &cached_roots {
+            self.flush_slot_cache(*root);
+            self.accounts_cache.set_max_flush_root(*root);
+        }
+
+        // Only add to the uncleaned roots set *after* we've flushed the previous roots,
+        // so that clean will actually be able to clean the slots.
+        self.accounts_index.add_uncleaned_roots(cached_roots);
+        flush_roots_elapsed.stop();
+
+        // Note we don't purge unrooted slots here because there may be ongoing scans/references
+        // for those slot, let the Bank::drop() implementation do cleanup instead on dead
+        // banks
+
+        // If there are > MAX_CACHE_SLOTS, then flush the excess ones to storage
+        let old_slots = self.accounts_cache.find_older_frozen_slots(MAX_CACHE_SLOTS);
+        let total_excess_slot_count = old_slots.len();
+        let mut unflushable_unrooted_slot_count = 0;
+        let max_flushed_root = self.accounts_cache.fetch_max_flush_root();
+        for old_slot in old_slots {
+            // Don't flush slots that are known to be unrooted
+            if old_slot > max_flushed_root {
+                self.flush_slot_cache(old_slot);
+            } else {
+                unflushable_unrooted_slot_count += 1;
+            }
+        }
+
+        datapoint_info!(
+            "accounts_db-cache-limit-slots",
+            ("total_excess_slot_count", total_excess_slot_count, i64),
+            (
+                "unflushable_unrooted_slot_count",
+                unflushable_unrooted_slot_count,
+                i64
+            ),
+        );
+
+        // Flush a random slot out after every force flush to catch any inconsistencies
+        // between cache and written state (i.e. should cause a hash mismatch between validators
+        // that flush and don't flush if such a bug exists).
+        let num_slots_remaining = self.accounts_cache.num_slots();
+        if force_flush && num_slots_remaining >= FLUSH_CACHE_RANDOM_THRESHOLD {
+            // Don't flush slots that are known to be unrooted
+            let mut frozen_slots = self.accounts_cache.find_older_frozen_slots(0);
+            frozen_slots.retain(|s| *s > max_flushed_root);
+            // Remove a random index 0 <= i < `frozen_slots.len()`
+            let rand_slot = frozen_slots.choose(&mut thread_rng());
+            if let Some(rand_slot) = rand_slot {
```

Comment:
cool, it seems that we fixed out of bound error by starting to use `.choose()`. :)

Reply:
this could be possible, if the retained frozen_slots is empty.

Reply:
hehe yeah that was a good fix 😄